neverbleed.c

```diff
@@ -505,10 +527,57 @@ void neverbleed_transaction_write(neverbleed_t *nb, neverbleed_iobuf_t *buf)
     iobuf_transaction_write(buf, thdata);
 }
 
+static void do_exdata_free_callback(void *parent, void *ptr, CRYPTO_EX_DATA *ad, int idx, long argl, void *argp)
+{
+    assert(ptr != NULL);
+    struct st_neverbleed_rsa_exdata_t *exdata = ptr;
+    struct st_neverbleed_thread_data_t *thdata = get_thread_data(exdata->nb);
+
+    neverbleed_iobuf_t buf = {NULL};
+    iobuf_push_str(&buf, "del_pkey");
+    iobuf_push_num(&buf, exdata->key_index);
+    // "del_pkey" command is fire-and-forget, it cannot fail, so doesn't have a response
+    iobuf_transaction_no_response(&buf, thdata);
+
+    free(exdata);
+}
+
+static int get_rsa_exdata_idx(void);
+static void rsa_exdata_free_callback(void *parent, void *ptr, CRYPTO_EX_DATA *ad, int idx, long argl, void *argp)
+{
+    assert(idx == get_rsa_exdata_idx());
+    do_exdata_free_callback(parent, ptr, ad, idx, argl, argp);
+}
+
+static int get_rsa_exdata_idx(void)
+{
+    static volatile int index;
+    NEVERBLEED_MULTITHREAD_ONCE({
+        index = RSA_get_ex_new_index(0, NULL, NULL, NULL, rsa_exdata_free_callback);
+    });
+    return index;
+}
+
+static int get_ecdsa_exdata_idx(void);
+static void ecdsa_exdata_free_callback(void *parent, void *ptr, CRYPTO_EX_DATA *ad, int idx, long argl, void *argp)
+{
+    assert(idx == get_ecdsa_exdata_idx());
+    do_exdata_free_callback(parent, ptr, ad, idx, argl, argp);
+}
+
+static int get_ecdsa_exdata_idx(void)
+{
+    static volatile int index;
+    NEVERBLEED_MULTITHREAD_ONCE({
+        index = EC_KEY_get_ex_new_index(0, NULL, NULL, NULL, ecdsa_exdata_free_callback);
```

Comment:
Use of elliptic-curve functions should be guarded by `#ifdef NEVERBLEED_ECDSA`. I think we should either add #ifdef around `get_ecdsa_exdata_idx` and `ecdsa_exdata_free_callback`, or move the two functions inside the #ifdef that defines the rest of ECDSA-related functions, preferably directly before `ecdsa_get_privsep_data` so that the layout of the functions will mirror that of RSA.